File: DAIDALUS/Java/src/gov/nasa/larcfm/ACCoRD/OwnshipState.java
/*
 * Copyright (c) 2015 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
package gov.nasa.larcfm.ACCoRD;

import gov.nasa.larcfm.Util.EuclideanProjection;
import gov.nasa.larcfm.Util.Position;
import gov.nasa.larcfm.Util.Projection;
import gov.nasa.larcfm.Util.Vect3;
import gov.nasa.larcfm.Util.Velocity;
import gov.nasa.larcfm.Util.f;

public class OwnshipState extends TrafficState {
  
  private final EuclideanProjection eprj; // Projection from pos
  private final Vect3    s; // Projected position
  private final Velocity v; // Projected velocity

  public OwnshipState() {
    super();
    eprj = Projection.createProjection(Position.ZERO_LL);
    s = Vect3.INVALID;
    v = Velocity.INVALID;
  }
  
  public static final OwnshipState INVALID = new OwnshipState();
  
  public OwnshipState(String id, Position po, Velocity vo) {
    super(id,po,vo);
    if (pos.isLatLon()) {
      eprj = Projection.createProjection(pos.lla().zeroAlt());
      s = eprj.project(pos);
      v = eprj.projectVelocity(pos,vel);
    } else {
      eprj = Projection.createProjection(Position.ZERO_LL);
      s = pos.point();
      v = vel;
    }
  }
  
  public OwnshipState(TrafficState own) {
    this(own.id,own.pos,own.vel);
  }
  
  public Velocity inverseVelocity(Velocity v) {
    return eprj.inverseVelocity(s,v,true);
  }
    
  public Vect3 get_s() {
   return s;
  }
  
  public Vect3 pos_to_s(Position p) {
    if (p.isLatLon()) {
      if (!pos.isLatLon()) {
        f.pln("OwnshipState.get_proj_s() ERROR: ownship is Euclidean, but projecting a latlon postition");
        Thread.dumpStack();        
        return Vect3.INVALID;
      }
      return eprj.project(p);
    } 
    return p.point();
  }
  
  public Vect3 traffic_s(TrafficState ac) {
    return pos_to_s(ac.getPosition());
  }
  
  public Velocity get_v() {
    return v;
  }

  public Velocity vel_to_v(Position p,Velocity v) {
    if (p.isLatLon()) {
      if (!pos.isLatLon()) {
        f.pln("OwnshipState.get_proj_v() ERROR: ownship is Euclidean, but projecting from a latlon postition");
        Thread.dumpStack();        
        return Velocity.INVALID;
      }     
      return eprj.projectVelocity(p,v);
    } 
    return v;
  }

  public Velocity traffic_v(TrafficState ac) {
    return vel_to_v(ac.getPosition(),ac.getVelocity());
  }
  
  public OwnshipState linearProjection(double offset) {
	  return new OwnshipState(id, pos.linear(vel, offset), vel);
  }

  public String toPVS(int prec) {
    return toPVS(id,s,v,prec);
  }
  
  public String toPVS(TrafficState ac, int prec) {
    return toPVS(ac.id,traffic_s(ac),traffic_v(ac),prec);
  }
  
}

File: DAIDALUS/C++/src/OwnshipState.cpp
/*
 * Copyright (c) 2015 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
#include "OwnshipState.h"
#include "Position.h"
#include "Velocity.h"
#include "Projection.h"
#include "format.h"

namespace larcfm {

  OwnshipState::OwnshipState() : TrafficState() {
    eprj = Projection::createProjection(Position::ZERO_LL());
    s = Vect3::INVALID();
    v = Velocity::INVALIDV();
  }

 OwnshipState::OwnshipState(const std::string& id, const Position& po, const Velocity& vo) : TrafficState(id,po,vo) {
    if (pos.isLatLon()) {
      eprj = Projection::createProjection(pos.lla().zeroAlt());
      s = eprj.project(pos);
      v = eprj.projectVelocity(pos,vel);
    } else {
      eprj = Projection::createProjection(Position::ZERO_LL());
      s = pos.point();
      v = vel;
    }
  }
  
  OwnshipState::OwnshipState(const TrafficState& own) : TrafficState(own.getId(),own.getPosition(),own.getVelocity()) {
    if (pos.isLatLon()) {
      eprj = Projection::createProjection(pos.lla().zeroAlt());
      s = eprj.project(pos);
      v = eprj.projectVelocity(pos,vel);
    } else {
      eprj = Projection::createProjection(Position::ZERO_LL());
      s = pos.point();
      v = vel;
    }

  }

  const OwnshipState OwnshipState::INVALID = OwnshipState();
  
  Velocity OwnshipState::inverseVelocity(const Velocity& v) const {
    return eprj.inverseVelocity(s,v,true);
  }
    
  Vect3 OwnshipState::get_s() const {
   return s;
  }
  
  Vect3 OwnshipState::pos_to_s(const Position& p) const {
    if (p.isLatLon()) {
      if (!pos.isLatLon()) {
        fpln("OwnshipState.get_proj_s() ERROR: ownship is Euclidean, but projecting a latlon postition");
        return Vect3::INVALID();
      }
      return eprj.project(p);
    } 
    return p.point();
  }
  
  Velocity OwnshipState::get_v() const {
    return v;
  }

  Velocity OwnshipState::vel_to_v(const Position& p, const Velocity& v) const {
    if (p.isLatLon()) {
      if (!pos.isLatLon()) {
        fpln("OwnshipState.get_proj_v() ERROR: ownship is Euclidean, but projecting from a latlon postition");
        return Velocity::INVALIDV();
      }     
      return eprj.projectVelocity(p,v);
    } 
    return v;
  }

  OwnshipState OwnshipState::linearProjectionOwn(double offset) const {
    return OwnshipState(id, pos.linear(vel,offset), vel);
  }

  Vect3 OwnshipState::traffic_s(const TrafficState& ac) const {
    return pos_to_s(ac.getPosition());
  }

  Velocity OwnshipState::traffic_v(const TrafficState& ac) const {
    return vel_to_v(ac.getPosition(),ac.getVelocity());
  }

  std::string OwnshipState::toPVS(int prec) const {
    return TrafficState::toPVS(id,s,v,prec);
  }

  std::string OwnshipState::toPVS(const TrafficState& ac, int prec) const {
    return TrafficState::toPVS(ac.getId(),traffic_s(ac),traffic_v(ac),prec);
  }

}
